pclntab/pclntab.go

```diff
@@ -0,0 +1,652 @@
+			// Note that `SearchGoPclntab` returns a slice starting from gopcltab header to the end of segment
+			// containing gopclntab. Therefore this slice might contain additional data after gopclntab.
+			// There does not seem to be an easy way to get the end of gopclntab segment without parsing the
+			// gopclntab itself.
+			if data, goPCLnTabAddr, err = SearchGoPclntab(ef); err != nil {
+				return nil, fmt.Errorf("failed to search .gopclntab: %w", err)
+			}
+			// Truncate the data to the end of the section containing gopclntab.
+			if sec := sectionContaining(ef, goPCLnTabAddr); sec != nil {
+				data = data[:sec.Addr+sec.Size-goPCLnTabAddr]
+			}
+			goPCLnTabEndKnown = false
+		} else {
+			var start, end libpf.SymbolValue
+			start, err = symtab.LookupSymbolAddress("runtime.pclntab")
+			if err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			end, err = symtab.LookupSymbolAddress("runtime.epclntab")
+			if err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			if start >= end {
+				return nil, fmt.Errorf("invalid .gopclntab symbols: %v-%v", start, end)
+			}
+			data = make([]byte, end-start)
+			if _, err = ef.ReadVirtualMemory(data, int64(start)); err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			goPCLnTabAddr = uint64(start)
+		}
+	}
+
+	if data == nil {
+		return nil, nil
+	}
+
+	goPCLnTabInfo, err = parseGoPCLnTab(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse .gopclntab: %w", err)
+	}
+
+	goPCLnTabInfo.Address = goPCLnTabAddr
+
+	if !goPCLnTabEndKnown && goPCLnTabInfo.Version >= ver116 {
+		funcDataSize := goPCLnTabInfo.findFuncDataSize()
+		if funcDataSize != -1 {
+			goPCLnTabSize := int(goPCLnTabInfo.Offsets.FuncTabOffset) + funcDataSize
+			if goPCLnTabSize < len(goPCLnTabInfo.Data) {
+				goPCLnTabInfo.Data = goPCLnTabInfo.Data[:goPCLnTabSize]
+			}
+		}
+	}
+
+	// Only search for goFunc if the version is 1.18 or later and heuristic search is enabled.
+	if goPCLnTabInfo.Version >= ver118 {
+		if symtab == nil {
+			symtab, _ = ef.ReadSymbols()
+		}
+
+		goFuncData, goFuncAddr, err := FindGoFunc(ef, goPCLnTabInfo, symtab)
+		if err == nil {
+			goFuncEndFound := false
+			if symtab != nil {
+				// symbol runtime.gcbits.* follows goFunc, try to use it to determine the end of goFunc.
+				nextSymAddr, err := symtab.LookupSymbolAddress("runtime.gcbits.*")
+				if err == nil && uint64(nextSymAddr) > goFuncAddr {
+					dist := uint64(nextSymAddr) - goFuncAddr
+					if dist < uint64(len(goFuncData)) {
+						goFuncData = goFuncData[:dist]
+						goFuncEndFound = true
+					}
+				}
+			}
+
+			if !goFuncEndFound {
+				// Iterate over the functions to find the maximum offset of the inline tree.
+				maxInlineTreeOffset := goPCLnTabInfo.findMaxInlineTreeOffset()
+				// If the inline tree offset is found, truncate the goFunc data.
```

Comment:
Can this fail? If we can't find the offset will we fail to truncate?

Reply:
If it fails, then goFunc is now discarded